sdk/compiler/damlc/lib/DA/Cli/Damlc/Command/UpgradeCheck.hs

```diff
@@ -0,0 +1,194 @@
+fromCollect = ExceptT . runCollectErrs
+
+toCollect :: CheckM a -> CollectErrs IO [CheckingError] a
+toCollect = CollectErrs . runExceptT
+
+-- Read main & deps from a DAR path, annotate with name and version so that
+-- upgrades knows which types to check against which
+type Archive = (NormalizedFilePath, UpgradedPkgWithNameAndVersion, [UpgradedPkgWithNameAndVersion])
+
+readPathToArchive
+  :: NormalizedFilePath
+  -> CheckM Archive
+readPathToArchive path = do
+  ExtractedDar{edMain,edDalfs} <-
+    catchIOException
+      (CECantReadDar path . displayException)
+      (extractDar (fromNormalizedFilePath path))
+  (main, deps) <- fromCollect $ do
+      main <- toCollect $ decodeEntryWithUnitId path Archive.DecodeAsMain edMain
+      deps <- traverse (toCollect . decodeEntryWithUnitId path Archive.DecodeAsDependency) edDalfs
+      pure (main, deps)
+  pure (path, main, deps)
+    where
+      catchIOException :: (IOException -> CheckingError) -> IO a -> CheckM a
+      catchIOException mkErr io = withExceptT (pure . mkErr) $ ExceptT $ try io
+
+      decodeEntryWithUnitId
+        :: NormalizedFilePath -> Archive.DecodingMode -> ZipArchive.Entry
+        -> CheckM UpgradedPkgWithNameAndVersion
+      decodeEntryWithUnitId darPath decodeAs entry =
+        fromEither (CECantReadDalf darPath (ZipArchive.eRelativePath entry)) $ do
+            let bs = BSL.toStrict $ ZipArchive.fromEntry entry
+            (pkgId, pkg) <- Archive.decodeArchive decodeAs bs
+            let (pkgName, mbPkgVersion) = LF.safePackageMetadata pkg
+            pure (pkgId, pkg, pkgName, mbPkgVersion)
+
+topoSortPackagesM :: [Archive] -> CheckM [Archive]
+topoSortPackagesM packages =
+  fmap (map withoutIdAndPkg) $ fromEither mkErr $ topoSortPackages (map withIdAndPkg packages)
+    where
+      cyclePkgToDep (_, (pkgId, _, name, mbVersion), _) = (pkgId, name, mbVersion)
+      mkErr cyclePkgs =
+        CEDependencyCycle (map (cyclePkgToDep . withoutIdAndPkg) cyclePkgs)
+      withIdAndPkg x@(_path, (pkgId, pkg, _name, _version), _deps) = (pkgId, x, pkg)
+      withoutIdAndPkg (_pkgId, x, _pkg) = x
+
+checkPackageAgainstPastPackages :: (Archive, [Archive]) -> CheckM ()
+checkPackageAgainstPastPackages ((path, main, deps), pastPackages) = do
+  let (_mainId, mainPkg, mainName, mbVersion) = main
+  case splitPackageVersion id <$> mbVersion of
+    Nothing -> pure ()
+    Just (Left _) -> pure ()
+    Just (Right rawVersion) -> do
+      let pastPackageFilterVersion pred = flip mapMaybe pastPackages $ \case
+            (_, x@(_, pastPackage, name, mbVersion), y) -> do
+              guard (not (isUtilityPackage pastPackage))
+              guard (name == mainName)
+              case splitPackageVersion id <$> mbVersion of
+                Just (Right rawVersion) -> do
+                  guard (pred rawVersion)
+                  pure (rawVersion, (x, y))
+                _ -> Nothing
+      let ordFst = compare `on` fst
+      let pastPackageWithLowerVersion =
```

Comment:
Suggestion: call it something like `closestPastPackageWithLowerVersion` or `greatestSmallerPastPackage`. Something that indicates that it's not just the absolute lower version.

Reply:
Done, improved the names as `closestPastPackageWithLowerVersion` and `closestPastPackageWithHigherVersion`